Roman-Yarinski/trading
# Trading v1

* Add: required packages and settings

* Add: required interfaces and SwapHelper

* Add: chainlink package

* Add: TradingPlatform smart contract

* Add: WETH interface

* Add: mock contracts for tests

* Add: functions for interactions with contract

* Refactor: fix config

* Test: add tests for TradingPlatform contract

* Refactor: add comment

* Refactor: add comment for   getSwapHelper method

* Refactor: fix config for test with UniSwap V3 fork

* Add: contract and methods for interacting with UniSwap V3

* Tests: add helpers for tests

* Tests: add tests for interacting with UniSwap V3

* Tests: add tests for main trading logic

* Add: read methods and fix small  issues

* Refactor: add config for deploy

* Refactor: fix MockERC20 contract file name

* Add: deployments scripts

* Fix: use array for user orders

* Refactor: remove useless code

* Refactor: use only needed imports

* Refactor: fix require massage

* Refactor: fix interfaces with ERC20

* Tests: add tests for SwapHelper

* Add: mock ERC20 contract with name

* Add: DCA and TRAILING orders

* Tests: fix tests for new Order struct

* Deploy: to goerli test net

* Add: logic for DCA and TRAILING add geters and cancel methods

* Tets: add and fix tests for trading platform contract

* Add: Two-way linking of orders, execution resultTokenOut for orders,
status to orderInfo struct

* Test: fix tests for new order conditions

* Fix: bound orders add params to constructor
refactor some namings

* Test: fix tests for new conditions

* Fix: config and types for TradingPlatform new constructor params

* Refactor: Add interface and comments for TradingPlatform contract

* Refactor: add docgen and generate docs for contracts

* Docs: regenerate docs

* Fix: some fixes after testing add emit OrderCanceled for bound orders
add result calculation for last order execution DCA/TRAILING fix check for createOrder and checkOrder
add role for setProtocolFee

* Test: fix and refactor tests for new contract conditions

* Test: fix Order types

* Refactor: add ReentrancyGuard and fix some comments

* Fix:  ReentrancyGuard import path and nonReentrant at performUpkeep

* Add: compatibility with automation LB for checkUpkeep, custom swap slippage.
Fix: closing DCA adn TRAILING orders,
token transfers to user

* Add: AutomationController contract

* Test: add additional tests for TradingPlatform

* Add: deployment script for UpkeepController

* Refactor: skip UpkeepController  coverage

* Refactor: generate docs

* Refactor: add config to README

* Refactor: fix solhint config

* Refactor: fix linter issues

* Refactor: default.example config for test values

* Refactor: fix license for contracts to MIT

* Refactor: config

* Refactor: hardhat config

* Add:  execution addTokensToWhitelist at TradingPlatform deploy

* Refactor: clean tests

* Add: hardhat task for price correction

* Refactor: README file

* Refactor: imports

* Refactor: fix types issues

* Refactor: update packages

* Refactor: fix ethers type

* Refactor: clean comments

* Refactor: auto audit with slither and triage

* Deploy: contracts to Goerli  testnet
